src/transaction.js

```diff
@@ -270,30 +268,35 @@ class Transaction extends TransactionRequest {
    * });
    */
   commit(callback) {
-    const self = this;
     if (this.ended_) {
-      throw new Error('Transaction has already been ended.');
+      callback(new Error('Transaction has already been ended.'));
+      return;
     }
+
     const reqOpts = {
       mutations: this.queuedMutations_,
     };
+
     if (this.id) {
       reqOpts.transactionId = this.id;
     } else {
       reqOpts.singleUseTransaction = {
         readWrite: {},
       };
     }
+
     this.request(
       {
         client: 'SpannerClient',
         method: 'commit',
         reqOpts: reqOpts,
       },
-      function(err, resp) {
+      (err, resp) => {
         if (!err) {
```

Comment:
When `end()` is called thats indicates that this transaction is over and that it is safe to release the connection back into the pool. It is possible that a concurrent transaction will trigger an abort error, which in case we need to retry the transaction in its entirety. There's also a possibility that a non-retryable error will come back, at which point you might want to perform a rollback.